### ISSUE TITLE
Something happened with my kdbx (Bad master key?)

### ISSUE THREAD

**Issue by IlluminatiWave** [Kunzisoft/KeePassDX#1537]
**Describe the bug**

Keepass cannot open my database having entered the master password (I saved the master password in a clipboard).

**To Reproduce**

1. Open KeepassDX
2. Open the KDBX file
3. Typing Master key
4. "incorrect credentials

**Expected behavior**

Open normally

**KeePass Database**

 - Created with: KeepassDX 3.5.0 (probably, the database is after February of this year).
 - Version: 1
 - Location: Local micro sd
 - File provider (`content://` URI): Everything was managed within the application 
 - Size: 2.82kb
 - Contains attachment: no

**KeePassDX:**

 - Version: KeepassDX 3.5.0 (probably, the database is after February of this year).
 - Build: Free
 - Language: Spanish

**Android:**

 - Device: Xperia XZ Premium (maple G8141)
 - Version: Android 13 (Nusantara custom rom / rooted. The database was created in the Android 9 Stock version. )

**Additional context**

I used it in conjunction with Firefox focus. All passwords were generated by the application itself (I did the master password manually).

Since I wanted to make the jump to Android 13, I backed up the password and saved it to a clipboard manager (Clip Stack). The password was still working at those times.

It is my understanding that every time a password/account is added or removed. The database is automatically updated in the file itself. I didn't really have any advanced settings (like storing in the keystore), the most complicated thing I used was the biometric manager.

When I wanted to restore my database (I copied it from the pc because for some reason, the ROM doesn't read large capacity SD cards, and I refuse to format my SD ), it tells me that "failed to read credentials". I have entered the password from the clipboard and manually. Neither option has worked.

Just in case, I tried other passwords (which for some reason I didn't delete from clip stack/they are passwords I had copied from keepassDX itself due to firefox focus not having support with KeePassDX on Android 9), with the same result (it was obvious it wasn't going to work, but I had to try).

Bottom line. For some unknown reason, The database is inaccessible even with master key when changing the rom.

Some information that I don't know if it's useful

When I created the database, I saved it to the SD card, set the password and started putting stuff into it. I used the biometric manager to unlock. When I changed ROM (before), I only backed up the master key (because I knew the database was automatically saved to the SD card. Which I checked by creating a new database with the same master key and adding more or less the same amount of keys I had in the old one, which gave me a similar size to the inaccessible database).

**Comment by J-Jamet** [comment]
> It is my understanding that every time a password/account is added or removed. The database is automatically updated in the file itself.

Indeed.

> Bottom line. For some unknown reason, The database is inaccessible even with master key when changing the rom.

The reproduction procedure is too generic and it is not possible to reproduce your problem, which seems to be related to a bad file handling or a bad master key, and not a bug of KeePassDX.
I think your file was simply copied or modified incorrectly during the ROM change.

**Comment by IlluminatiWave** [comment]
> is too generic

Well, it's not like I did some kind of strange move to save/load the database either, it seems to be more of a problem inside the database because I tried to create a new database which I opened on another device successfully (using the same master key as the damaged db).

> I think your file was simply copied or modified incorrectly during the ROM change.

I don't think it was that because from the beginning the database was generated inside the SD card. The only thing I did was to copy my master key to Clip Stack and make a clip stack backup (hence I was able to "backup" a few accounts).

So the rom change I doubt it had anything to do with it because the db was always (and still is) in the same directory in the first place (in fact the one I'm trying to open is a copy which doesn't open at the moment).

I can only think of Android 9 having some specific encoding to show symbols like "Ç" or numbers as superscript ( ⁰¹²³⁴⁵⁶⁷⁸⁹⁹ ).

Most likely my only option is to crack my own kdbx 0_o

**Comment by J-Jamet** [comment]
I don't understand what you did with the clipboard. You have to know your master key to open your database, but if you save your master key on a flat file next to the base file, what is the point?

If you have made a copy of your database and you have the master key, you can open your database with another KeePass software, like KeePassXC if you think KeePassDX is involved. Normally, special characters are well recognized. If your main key does not work, there are two possible problems: 
 - your master key is not the right one
 - your file is corrupted
In the first case, you cannot open your database without the master key.
In the second case, you need a working backup.

**Comment by IlluminatiWave** [comment]
> I don't understand what you did with the clipboard

Clip stack has an option to make backup copies (let's say I'm a clipboard guy). Which I used because it has never given me failures.

> You have to know your master key to open your database

For convenience I saved it in the clipboard before (it is a phrase with special symbols like "pa$$w°rd", but it is possible to type it from the android keyboard. I didn't expect that the custom rom could not read the micro sd.)

> you can open your database with another KeePass software

Same error, so I discarded that it was the app. Although it is rare because the master key also works in desktop applications (the one with the corrupted database).

>> it seems to be more of a problem inside the database

> * your master key is not the right one

Maybe, hopefully that's it. I'll try my luck scratching in my clipboard history.

> * your file is corrupted

I tried to recover it on the computer from a new database, failure as expected.

I was thinking of other possible causes. For example that for some unknown reason I had a key file, which would be odd because I don't remember setting one up. In which case it is just a random thought. There is no way to know if a database needs a master key or not (which I consider a keepass problem really, it is not very transparent in that aspect). It may provide security to not know if a file needs a key file or not, but for "novice" users of password managers, it is a problem (and more if you are not sure if you must pass the same password with all the files on the drive by brute force).

**Comment by J-Jamet** [comment]
> There is no way to know if a database needs a master key or not

The main purpose is that the user has to remember his master credentials. If you use another software, it makes another failure point. The proof is that you don't even remember if you had a key file.

> and more if you are not sure if you must pass the same password with all the files on the drive by brute force.

I'm going to consider this a troll, it's obvious that the goal of KeePass is to avoid brute force as much as possible.

**Comment by IlluminatiWave** [comment]
> There is no way to know if a database needs a master key or not

This was my fault, I was referring to the attachment that is used in conjunction with the master key. There is no way of knowing or not

> purpose is that the user has to remember his master cr....

I know my master key, I just manually typed it into the clipboard because I didn't want to spend time manually typing it when recover my passwords. I already wrote the master key (manually), And it keeps giving error.

>  the goal of KeePass is to avoid brute force as much as possible.

If I know my own master key (Because I had to write it every time I wanted to access my database), the only "brute force" thing I'll use is my master key along with all the files on my sd card. Although there is also the possibility that it has been corrupted.

>>Máster key + file 1
>>Master key + file 2
>>Master key + file n

Does KeePassDX always modify the database? Or only when modifications are made to it. (For example, does KeePassDX keep the file intact if I just go in to "check" my keys?)

> consider this a troll

I'm not trying to troll. As far as I can remember, I never used a key file. But I do not rule out that possibility. So my only 2 options are that I have used a file, or that the db is damaged (I would have been able to recover the database with my master key from the desktop. But when trying it does not return anything. That's why I had my doubts if android 9 used some special encoding that you modified the utf code of the special characters)

**Comment by IlluminatiWave** [comment]
Well, after having done several tests, I think that my database is lost (possibly damaged at some point).

However, I noticed a few things (on the pc with KeePass 2.53.1 in CLI)

Windows in its console (BATCH files) uses ANSI text encoding, Which creates errors because if the password was UTF-8 (Pe¶ª ⁹), then Windows represents it as "Pe┬Â┬¬ Ôü╣", which creates errors to open the database (the text changes due to the encoding) this is solved by configuring the batch with a "chcp 65001" (utf-8 encoding).

I really don't know if this is a problem for Android versions with special encoding. For example, I understand that for Japan shift-jis encoding is used (in Windows it is "chcp 932", which represents the example password as "Peﾂｶﾂｪ 竅ｹ" ), when changing region/encoding, it "could" mark the password as incorrect due to these changes in the encoding (I don't know if this has solution forcing the utf-8 encoding in keepasdx).

I have seen this encoding problem before when someone tries to run Japanese-made programs (shift-jis) on "western" machines (utf-8). All this talking about windows, I have no experience with japanese programs for android.

----------------------

This is no longer related to the problem itself, but I will share what I did to create my "brute force" batch.

1. From the pc (windows) open the console and type cmd (in case powershell opens).
2. change the encoding of the console with the command "chcp 65001" (this can solve problems when encoding the list)
3. select the location of the folder with the possible key files (in my case it was all my external drive "cd D:")
4. Create a list with all files except folders with "dir /B/S/A-d > list.txt" (the command "/B/S" is to show the directories while "/A-d" excludes directories and ">" is to save the list in a file).
5. Opening the file (list.txt) in notepad++

Referring to the documentation of keepas cli, the script should have the following format

`"Path\KeePass.exe" "dbpath\MyDatabase.kdbx" -pw: "password" -keyfile: "kfpath\KeyFile".`

6. open the replace window, select regular expression and

find= $ (this selects the end of line)
replace= " (this is to add quotation marks at the end of each line)

7. Add the script to be executed at the beginning of each line replacing the same beginning of all lines. As in my case it was the whole "D:drive", I only have to adapt the following command:

`START /W "Path\KeePass.exe" "dbpath\MyDatabase.kdbx" -pw: "password" -keyfile: "D:\".`

In this way, all the lines go from 

```
D:\file_1.jpg
D:\file_2.mp4
D:\folder_nfile_n.xyz
```

to

```
START /W "Path\KeePass.exe" "dbpath\MyDatabase.kdbx" -pw: "password" -keyfile: "D:\file_1.jpg"
START /W "Path\KeePass.exe" "dbpath\MyDatabase.kdbx" -pw: "password" -keyfile: "D:\file_2.mp4"
START /W "Path\KeePass.exe" "dbpath\MyDatabase.kdbx" -pw: "password" -keyfile: "D:\folder_n\file_n.xyz"
```

the "START /W" command says to run keepass and wait until you close the program to execute the next command, so the process is semi-automatic (the user just has to close the program manually, as there is no way to know if the database was opened correctly from the CLI). 

8. Add chcp 65001 at the beginning of the file so that the password and directories are read in utf-8 and not ANSI.

9. Save the file and rename it to list.bat (txt > bat).

10. Execute the .bat file and close all keepas windows that give error until one opens the database successfully.

I did not add the command "@echo off" because this hides the command that is currently running, which is useful in case the database was opened and you closed the window by mistake (in which case you should only memorize the last 2 or 3 files that were opened, one of those should be the keyfile).
